### DISCUSSION TITLE
Support for double/decimal values

### DISCUSSION THREAD

**Discussion by lipchev** [angularsen/UnitsNet#875] [Ideas]
I think we've largely covered the _generics_ route for doing this in #714. 
But have you considered boxing the Quantity.Value with something like the QuantityValue that is used in the constructor? It seems to me that we'd only need to implement [IConvertible](https://docs.microsoft.com/en-us/dotnet/api/system.iconvertible?view=net-5.0), add a few operators like `QuantityValue + QuantityValue, QuantityValue + double, QuantityValue + decimal` and possibly some constructor supplied strategy for determining the resulting type of a mixed type operation. 
Example usage:
```
QuantityValue value = Mass.FromGrams(1m).Milligrams; // contains decimal
double actualValue = value; // that's doing Decimal.ToDouble
decimal mayOverflow = (decimal)value; // option A: explicit casting required (may overflow with default strategy)
decimal maxDecimal = (decimal) new QuantityValue(double.MaxValue, TypeStrategy.HandleOverflows)) // option A with custom strategy for handling overflows by using Decimal.Min/Max
decimal possibleInformationLoss = value; // option B (ambiguous when using var): safe implicit casting- assuming HandleOverflows is always true
Mass resul1 = Mass.FromGrams(1m) + Mass.FromGrams(1m); // contains decimal
Mass result2 = Mass.FromGrams(1m) + Mass.FromGrams(1d);  // contains decimal (default strategy -> first parameter type)
Mass result3 = Mass.FromGrams(1d) + Mass.FromGrams(1m, TypeStrategy.Preserve);  // contains decimal (default strategy + decimal strategy = decimal strategy)
Ratio positiveInfinity = Mass.FromGrams(1d, TypeStrategy.HandleOverflows) / Mass.FromGrams(0m, TypeStrategy.Preserve);  // contains QuantityValue.PositiveInfinity which converts to Double.PositiveInfinity or Decimal.MassValue
Ratio divisionByZeroException = Mass.FromGrams(1d) / Mass.FromGrams(0d);  // throws DivisionByZero (currently throws ArgumentException)
```

I'm not yet sure what the TypeStrategy (or whatever the name might be) would have as options but something along the lines of:

- Auto: If both operands are Auto- the final type determined by the order of parameters
- Preserve: If the strategy for the first operand is Auto, the type of the second operand is used
- HandleOverflows (this could either be an added flag, or maybe a second Enum)- in practice I don't think anybody that would care enough to use the decimals would end up mixing them in operations with doubles, and it seems even less likely that an actual conversion overflow would ever occur- _but_, having the possibility of avoiding the DivisionByZeroException (by using the dedicated QuantityValue.NaN/Infinity values) is something that I feel could be quite useful ([here](https://github.com/angularsen/UnitsNet/issues/848#issuecomment-749008876) is my typical use case).. I would even set it globally in my AppInit if I could..

**Comment by angularsen** [comment]
Clever idea, although it does feel a bit complicated at first glance.

My initial thinking is that we might be better off letting the user choose a numeric type across the board, similar to Math.NET matrices, but I do see the benefit of allowing any numeric type and giving the option to control overflows/infinity.

I'll let the idea mature a bit in my head.

**Comment by lipchev** [comment]
CC @tmilnthorp

**Comment by tmilnthorp** [comment]
I had thought of something similar, albeit constructing with an ```IConvertible```. Of course it's not CLS compliant, so that went out the window.

I don't think this is a bad strategy. Does it have any pros/cons over switching to generics at some point?

**Comment by angularsen** [comment]
This was also independently raised here: https://github.com/angularsen/UnitsNet/discussions/1048

> As for using QuantityValue as the internal value, maybe. I haven't really thought about it, but maybe there is some merit to it. I'm trying to come up with counter-arguments not to do so, and I think the best I can think of right now is that serialization will be more complicated and it probably adds some extra bytes for memory sensitive applications.

The best argument I could think of against might be serialization, but perhaps if we stick with double/decimal and not `Complex` as discussed there, then maybe serialization is unaffected?

I think `QuantityValue` instead of `double` may solve most of the gripes we have with double vs decimal.

I don't see generics happening anytime soon. I think it's a huge breaking change to go from value types to reference types and also I think value types is the better choice.

We could keep `double` / `decimal` as-is internally, but expose it differently via the interface, so you get `QuantityValue` from `IQuantity.Value` and `decimal` or `double` from concrete implementations like `Information` and `Length`:

```cs
Information oneByte1 = Information.FromBytes(1);
IQuantity oneByte2 = (IQuantity)oneByte1;

var bytes1 = oneByte1.Value; // decimal
var bytes2 = oneByte2.Value; // QuantityValue
```

**Reply by angularsen** [reply]
By the way, changing the return type of public interfaces would be a breaking change and candidate for #982 even if they in most cases probably would compile just fine due to implicit casts. However, at least reflection would fail and possibly some other cases I can't think of.

**Reply by pgrawehr** [reply]
`double` doesn't implicitly cast to `decimal` (nor the other way round). 

One can implicitly cast double to `QuantityValue`, but also not the other way round (which is good, as it would cause ambiguities IMHO). In this case, we would need to make sure we're not loosing precision when casting decimal to QuantityValue. 

So this would definitely be a breaking change and would need to go to V5.

**Reply by angularsen** [reply]
You are right, implict casts only go into `QuantityValue`, while explicit cast is required for getting values out. If we were to use this type as the return value, I would add `DoubleValue` and `DecimalValue` properties wrapping the explicit cast, to make it more discoverable.

**Reply by pgrawehr** [reply]
Sounds reasonable, even though it would require another indirection to get the actual data: 
`Console.WriteLine($"It's {myVar.Kilometers.DecimalValue} kilometers to go");` 

Since V5 has the printing improved, that should partially account for this ;-)

**Reply by angularsen** [reply]
To be clear, I only intended for `IQuantity.Value` to return `QuantityValue`, and instead let `Information.Value` return `decimal` directly.

So `myVar.Kilometers` would be `double` in this case assuming it was statically known as type `Length`, while `myIQuantityVar.Value` would require invoking `DecimalValue` or `DoubleValue`. 

We also have `IDecimalQuantity` to make this even more fun, where it could in fact return `decimal` also.

**Reply by pgrawehr** [reply]
This is getting confusing... I guess we need some kind of prototype to see how this all works out.

**Comment by tmilnthorp** [comment]
Generics don’t require reference types btw

**Reply by angularsen** [reply]
Ah, I keep mixing that one up. This is like the 3rd time. Do you recall the summary of previous generics discussions? What are pros/cons of generics vs QuantityValue?

**Reply by angularsen** [reply]
At first glance, I would really like to avoid having to change existing code from `Length.FromMeters(1)` to `Length<double>.FromMeters(1)`. I guess we could work around that by letting `Length` extend `Length<double>`, and similarly `Information` extend `Information<decimal>`.

**Reply by angularsen** [reply]
`IQuantityValue.Value` could still return `QuantityValue`, as discussed above. This might be a good compromise of the two solutions.

**Reply by tmilnthorp** [reply]
> I guess we could work around that by letting Length extend Length<double>, and similarly Information extend Information<decimal>.

If only structs allowed derivation 😞 

And C# doesn't have a type alias where we could say ```typedef Length = Length<double>```

**Comment by lipchev** [comment]
One (big) problem was how to make them work as single nuget, or alternatively how to make the two nugets work with one another.

**Reply by angularsen** [reply]
It will be the same nuget as before, no?

It just adds twice the quantity types:
`Length` + `Length<T>`
`Information` + `Information<T>`

And as suggested above, maybe let `Length` extend `Length<double>` and `Information` extend `Information<decimal>` so people can keep their code pretty much as-is.

**Reply by lipchev** [reply]
I was talking about to cons for the generics- we were discussing exposing something like UnitsNet.Double or UnitsNet.Decimal - having the concrete type in its own namespace and so on..

We cannot have both `Length` and `Length<T>` unless we double the fun (structs). However - we could have `IQuantity<T>`.

**Reply by tmilnthorp** [reply]
> One (big) problem was how to make them work as single nuget, or alternatively how to make the two nugets work with one another.

Or we have a separate NuGet: ```UnitsNet Generics``` 🤔

**Reply by angularsen** [reply]
> If only structs allowed derivation 😞

Aha, that was my confusion. Structs support generics, but not inheritance. Remember, brain, remember!

Isn't that a blocker for generics though? I really don't see multiple nugets or having to use `Length<double>` syntax everywhere.

Either, we switch to reference types (class or records), which makes the generics path seem more feasible by allowing `Length` extending `Length<double>` etc.

Or, we stick with struct value types, and maybe let `IQuantity.Value` return `QuantityValue` instead of `double`

**Reply by tmilnthorp** [reply]
> Isn't that a blocker for generics though? I really don't see multiple nugets or having to use Length<double> syntax everywhere.

I don't see it as a blocker. I would expect to see ```Length<double>``` if we wanted to allow for an explicit underlying type, just as people are used to seeing ```IEnumerable<int>```. 

> Either, we switch to reference types (class or records), which makes the generics path seem more feasible by allowing Length extending Length<double> etc.

Struct vs class: I don't mind either way. You can make a class immutable so that it "works" like a struct.

**Reply by pgrawehr** [reply]
Our types are generally small (only 2-3 fields), so there could be a performance penalty when changing to class. As one typically has a large number of instances of quantities, this could mean a lot more work for the GC. But this would need some analysis. 

It is a fact though, that classes are much more flexible.

**Reply by angularsen** [reply]
### Rant section
I don't think the comparison with `IEnumerable<T>` is fair, because the alternative would be `IEnumerable` and boxing items as `object`. Lists inherently need to support a vast range of item types.

Units.NET quantities at best need to support a couple of numeric types, and I am not convinced this is a common scenario. It seems nice to have, but my hunch is that very few will actually use it. Maybe on embedded, `float` may be favored, or certain domains preferring `decimal` perhaps.

To put it into perspective, **only 3 out of 112 quantities** use `decimal`; `Information`, `BitRate`, `Power`.

There is a lot of brain-pain and lengthy discussions just because of those 3 quantities. I'm being cheeky, because yes, I would also love to support any numeric type, but I can't really remember anyone complaining about the lack of decimal or float - they have usually come asking about why some are double and some are decimal, because it isn't intuitive at all.

Just a bit of history; for years it was all `double`, then `Information` came a long and it seemed like a good idea at the time to avoid rounding errors on bits. Same with `BitRate`. Then `Power` got some very large units and the `ConversionRoundTrip` test started failing due to precision errors. This was probably a mistake that has since been fixed, because I just tried to change `Power` back to `double` and tests are still green with the required precision `1e-5`. So that leaves only 2 quantities that has some need for `decimal` today.

End of rant 😅

Solutions to problems below.

**Reply by angularsen** [reply]
## Problem statement, an attempt

1. Confusing that some quantities use `decimal` and others `double`
1. `IQuantity.Value` loses precision for the 3 `decimal` based quantities
1. It would be nice if the numeric type could be chosen per quantity

## Solutions

A - IQuantity.Value returns QuantityValue
B - Use QuantityValue as internal value representation
C - Generics with structs
D - Generics with class/record types

### A - IQuantity.Value returns QuantityValue
Solves problem 2.

Pros:
- Quantities stay value types (immutable, can't be null, no GC pressure on allocating instances)
- The memory footprint remains the same (double or decimal, not both)

Cons:
- Some quantities will still be decimal while others are double

### B - Use QuantityValue as internal value representation

Solves problems 2 and 3.

Pros:
- `QuantityValue` could possibly be extended to support `float`, `Complex` etc.

Cons:
- No type-safety on the numeric type, so if you call `.ValueDecimal` and the quantity was constructed with `double`, you would risk loss of precision.
- Serialization is more complicated, more numeric types to implement support for and breaks backwards compatibility.
- Memory footprint increases with the size of `QuantityValue` (double 8 + decimal 16 + float 4 = 28 bytes). Then [I also learned](https://stackoverflow.com/a/1381327/134761) that `Nullable<T>` typically adds 4/8 bytes on a 32/64-bit system due to memory alignment, so we're probably looking at 40-50 bytes per quantity. Starting to get chonky. Structs are generally advised to [be less than 16 bytes](https://stackoverflow.com/a/2407743/134761).

### C - Generics with structs
Solves problems 1, 2 and 3.

Pros:
- Generics allows user to choose their preferred numeric type

Cons:
- All consumers must migrate to `Length<double>` syntax, since structs can't inherit
- Serialization must support generic values

### D - Generics with class/record types
Solves problems 1, 2 and 3.

Pros:
- Generics allows user to choose their preferred numeric type
- Syntax is "backwards compatible", if `Length` extends `Length<double>`

Cons:
- Losing value semantics (not-nullness, but immutability we can preserve)
- Serialization must support generic values
- Increased GC pressure might be an issue for some.
- Record types is only available in netstandard [with some hacks](https://sergiopedri.medium.com/enabling-and-using-c-9-features-on-older-and-unsupported-runtimes-ce384d8debb), but we could probably multi-target to use records only for .NET5+ and use classes for netstandard2.

**Reply by angularsen** [reply]
Please help refine this list.

After typing it all out, I am currently leaning towards combining these two solutions:
- A - IQuantity.Value returns QuantityValue
- D - Generics with class/record types

Solution A seems fine, but isolated it only fixes a minor problem of rounding errors.
Solution B does not scale well with many numeric types, structs become fat.
Solution C has a cumbersome syntax and is a huge breaking change for all consumers.
Solution D is perhaps the better solution, if we agree that changing from value types to reference types is fine.

**Reply by angularsen** [reply]
@josesimoes you are doing some embedded work I assume.
Since nanoFramework has its own code generator, it can choose its own path or stay as-is.

However, do you have any thoughts on the above? In summary:
- Should we let users choose the numeric type?
- Should quantities be value types or reference types, or does it not matter much?

**Reply by josesimoes** [reply]
At nanoFramework (and because it's usage context: microcontrollers) we usually go with float and, in case extra precision is required doubles. We don't even have native support for decimals. It simply doesn't make sense, not to mention that the memory usage it's overkilling: 128 bits vs 64 bits for double. 
Unless one is dealing with astronomic or sub-atomic quantities, there's no real use for it.

Having said all that 😅 , I always advocate for the simplest and effective approach to solve a problem. I'm not seeing any added value on bringing generics to UnitsNet and even less to allow users choose the numeric type. That can be done, of course. It might look sexy or trendy, but, at the end of day, what's the real benefit from that? I don't see any, to be honest.
It will take a lot of refactoring, possibly will introduce breaking changes that will force folks to rework their code, and so on. Because of? 
Just because we can do something, it doesn't mean we should do it. 😉 
My take on this is that a library such as this one, should be simple and effective. Do it with a pragmatic approach. That possibly means choosing the underlying numeric types and that's it. The value of the library comes from the ease of use it offers with all these units. Not from the fact that's using decimal or double types.
This may be controversial, I agree, and it may provoke passionate discussions, I get that too. I keep thinking that it should be kept simple and effective.

Answering your last question: on quick though I don't think it matters _that_ much if quantities are value types or reference ones... 

Another heavy consumer of the library are the [.NET IoT bindings](https://github.com/dotnet/iot) maybe @Ellerbach should add his take on this too.

**Reply by Ellerbach** [reply]
Ping @pgrawehr @joperezr with whom we've discussed yesterday this specific element.

Note that on the nanoFramework side, we can always replace during the code generation decimal by double. Not ideal but that can be done with a warning on the limitations.

**Reply by angularsen** [reply]
Also, nanoFramework has its entirely own code generator, so it does not have make use of generics if the main library choose to do so.

**Reply by pgrawehr** [reply]
> IQuantity.Value loses precision for the 3 decimal based quantities 

That's only half the problem. Worse (at least for me now) is that all the conversion properties also return double. 

Generally, I think going with D is the only option that allows for future flexibility, if we do anything on this list at all. It does have the downside of added GC pressure, but I guess we would need to have some test case to demonstrate when this would be a real problem. 

So I agree, the combination of A + D is conceptually the best.

**Reply by pgrawehr** [reply]
One thing to consider, though: Using generics would require the use of the new generic math interfaces. These are only in preview right now.

**Reply by tmilnthorp** [reply]
> One thing to consider, though: Using generics would require the use of the new generic math interfaces. These are only in preview right now.

And won't be netstandard compliant 😢

**Reply by angularsen** [reply]
Good point about arithmetic, however generic arithmetic is possible without those interfaces, just less convenient.
Some initial work and discussion here: #698

I agree with the take of @josesimoes , that we risk implementing a fancy generic quantity system that complicates everything and only benefits a small minority of the user base. The mission of Units.NET has from the start been to simplify things for most users, not to cover every use case.

### Proposal
- Identify 1-3 persons who really wants this feature and become the champions of this feature. Anyone here?
- Build a prototype
- Publish a separate nuget `UnitsNet.Xp.Generics` or similar, to indicate it is experimental and not a primary package

We can then see if the idea gets any traction, both on implementing it and on user adoption.
We can also much faster iterate on the idea without having to think about breaking people's code, and get a better feel of pros and cons.

If it turns out this is something a lot of people have been silently waiting for, or we simply find the improvement worth it ourselves, then it's much easier to bring into the main library later.

What do you guys think?

**Reply by angularsen** [reply]
Solution `A - IQuantity.Value returns QuantityValue` seems like a good idea regardless, I think.

Also @pgrawehr wrote:
> That's only half the problem. Worse (at least for me now) is that all the conversion properties also return double.

Right, I had not thought of that. I see no reason we shouldn't change those to `decimal`. Would you be interested in attempting a PR?

```cs
UnitsNet.Information.FromKilobytes(5).Bytes; // double, should be decimal
```

**Reply by pgrawehr** [reply]
@angularsen I can try, but I can't say when I find time. Also, I guess that needs a change to the code generator, which I haven't looked into.

**Reply by angularsen** [reply]
@pgrawehr No problem, just let us know if you start on it by creating a PR early and then we can assist you there if you have any questions.

**Reply by pgrawehr** [reply]
@angularsen : I've started looking into this, but would you mind updating the v5 branch with the relevant changes from master? I get a lot of merge conflicts (not counting the generated code) of which I'm unsure how to resolve.

**Reply by angularsen** [reply]
@pgrawehr Great, I just merged in master and bumped the release branch to 5.0.0-alpha005.
The merge conflicts can be a pain for sure with all the generated code :-/

**Comment by lipchev** [comment]
I think a key point that we should probably also have a discussion about is what would be the ideal internal representation of the value field. Should it be double/decimal, QuantityValue or even maybe something fancier like [Fractions](https://github.com/danm-de/Fractions)?

**Reply by pgrawehr** [reply]
Well,  if we go with the Generic solution, it could be user-defined.

**Comment by angularsen** [comment]
For visibility, I am re-posting this: https://github.com/angularsen/UnitsNet/discussions/875#discussioncomment-2321700


Although generic value types seems nice to have, it is likely that only a minority will make use of it and it adds a lot of complexity to our code base. The mission of Units.NET has from the start been to simplify things for most users, not to cover every use case.

### Proposal

#### Implement solution A - `IQuantity.Value returns QuantityValue` in the main nuget
  - @pgrawehr is looking into it https://github.com/angularsen/UnitsNet/discussions/875#discussioncomment-2394796
  - Related, conversion properties like `Information.Bits` should also return `decimal` instead of `double`

#### Implement solution D - `Generics with class/record types` as a prototype nuget
  - Identify 1-3 persons who really wants this feature and become the champions of this feature. Anyone here?
  - Build a prototype
  - Publish a separate nuget `UnitsNet.Xp.Generics` or similar, to indicate it is experimental and not a primary package

We can then see if the idea gets any traction, both on implementing it and on user adoption.
We can also much faster iterate on the idea without having to think about breaking people's code, and get a better feel of pros and cons.

If it turns out this is something a lot of people have been silently waiting for, or we simply find the improvement worth it ourselves, then it's much easier to bring into the main library later.

What do you guys think?

**Comment by AndreasLeeb** [comment]
Regarding solution A:
If `QuantityValue` became the internal representation of the quantities, there would be an easy way to cut its size by 60%.

Taking advantage of the fact that `decimal` [does not use bits 0 to 15 (and also bits 24 to 30) of its fourth 32-bit portion](https://docs.microsoft.com/en-us/dotnet/api/system.decimal.getbits?view=netstandard-2.0#system-decimal-getbits(system-decimal)), we can construct a union struct that uses the first byte of this unused memory (meaning all zeroes) to save whether a `decimal` or `double` is stored (other types, such as `long`, `ulong`, `float`, `int` etc. could also be supported in this manner). This way, both fields overlap, and moreover, no additional memory for the nullability used to distinguish the underlying types in the current implementation is needed.

Small prototype with only rudimentary operator support and handling of different underlying types:

    [StructLayout(LayoutKind.Explicit, Size = 16)]
    public readonly struct UnionQuantityValue
    {
        public UnionQuantityValue(decimal value) : this()
        {
            Decimal = value;
            Type = DataType.Decimal; // not actually necessary due to bit structure of decimal
        }

        public UnionQuantityValue(double value) : this()
        {
            Double = value;
            Type = DataType.Double;
        }

        [FieldOffset(0)]
        public readonly decimal Decimal;

        [FieldOffset(8)]
        public readonly double Double;

        [FieldOffset(0)]
        public readonly DataType Type;

        public bool IsDecimal => Type == DataType.Decimal;

        public bool IsDouble => Type == DataType.Double;

        public static implicit operator decimal(UnionQuantityValue value) => value.Type switch
        {
            DataType.Decimal => value.Decimal,
            DataType.Double => (decimal)value.Double,
            _ => throw new NotImplementedException(),
        };

        public static implicit operator UnionQuantityValue(decimal @decimal) => new(@decimal);

        public static implicit operator double(UnionQuantityValue value) => value.Type switch
        {
            DataType.Decimal => (double)value.Decimal,
            DataType.Double => value.Double,
            _ => throw new NotImplementedException(),
        };

        public static implicit operator UnionQuantityValue(double @double) => new(@double);

        public static UnionQuantityValue operator +(UnionQuantityValue a, UnionQuantityValue b)
        {
            ConvertTypesIfNecessary(ref a, ref b);

            return a.Type switch
            {
                DataType.Decimal => a.Decimal + b.Decimal,
                DataType.Double => a.Double + b.Double,
                _ => throw new NotImplementedException()
            };
        }

        public static UnionQuantityValue operator -(UnionQuantityValue a)
            => a.Type switch
        {
            DataType.Decimal => -a.Decimal,
            DataType.Double => -a.Double,
            _ => throw new NotImplementedException(),
        };

        private static void ConvertTypesIfNecessary(ref UnionQuantityValue a, ref UnionQuantityValue b)
        {
            if (a.Type == b.Type)
            {
                return;
            }

            a = (decimal)a;
            b = (decimal)b;
        }

        public override string ToString()
            => Type switch
        {
            DataType.Decimal => Decimal.ToString(),
            DataType.Double => Double.ToString(),
            _ => throw new NotImplementedException(),
        };

        public enum DataType : byte
        {
            Decimal = 0,
            Double = 1
        }
    }

Size comparison

    Marshal.SizeOf<UnionQuantityValue>(): 16 B
    Marshal.SizeOf<UnitsNet.QuantityValue>(): 40 B

**Reply by angularsen** [reply]
Today I learned! This is awesome insight.

Some related work on v5 branch relating to `QuantityValue`: #1074 

The internal representation is not changed, so `Length` still stores `double` internally, and `Information` stores `decimal`. However, `IQuantity.Value` is now being changed from `double` to `QuantityValue`, in order to preserve the original value. The reduced size would benefit there.

I think what you are proposing is a win even if we don't change the internal representation.
Also I don't see any downsides to your idea, since you already covered how other numeric types can be supported too. The only problem would be numeric types too big to fix, such as [Complex](https://docs.microsoft.com/en-us/dotnet/api/system.numerics.complex?view=net-6.0) or various big number representations. We can revisit if that ever becomes a need.

Would you be interested in attempting a pull request against `release/v5` branch?

**Reply by AndreasLeeb** [reply]
#1084
